src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -82,33 +84,73 @@ cataNode f (Node pos ty subs) = f pos ty (cataNode f <$> subs)
 
 -- | Remove nodes with accordance with global `MarkdownConfig` and local
 --   overrides.
-removeIgnored :: Node -> Either Text Node
-removeIgnored = runIdentity . runExceptT . flip evalStateT None . cataNode remove
+removeIgnored :: MonadWriter [ScanError] m => FilePath -> Node -> m Node
+removeIgnored fp = withIgnoreMode . cataNode remove
   where
     remove
-      :: (MonadError Text m, MonadState IgnoreMode m)
+      :: (MonadWriter [ScanError] m, MonadState IgnoreMode m)
       => Maybe PosInfo
       -> NodeType
       -> [m Node]
       -> m Node
     remove pos ty subs = do
       mode <- get
+      let node = Node pos ty []
       case (mode, ty) of
         (Paragraph, PARAGRAPH) -> put None $> defNode
-        (Paragraph, x)         -> throwError (makeError mode (prettyType x) pos)
-        (File, _)              -> throwError (makeError mode "" pos)
-        (Link, LINK {})        -> put None $> defNode
-        (Link, _)              -> Node pos ty <$> sequence subs
-        (None, _) -> do
-          case getIgnoreMode (Node pos ty []) of
-            Just mode' -> put mode' $> defNode
+        (Paragraph, x)         -> do
+          tell . scanError pos . makeError mode $ prettyType x
+          put None
+          Node pos ty <$> sequence subs
+        (File, _)              -> do
+          tell . scanError pos $ makeError mode ""
+          put None
+          Node pos ty <$> sequence subs
+        (Link _, LINK {})      -> put None $> defNode
+        (Link pos', _)         ->
+          case getIgnoreMode node of
+            Just mode'  -> do
+              tell . scanError pos' $ makeError mode ""
+              handleMode node mode'
+            Nothing     -> Node pos ty <$> sequence subs
+        (None, _)              ->
+          case getIgnoreMode node of
+            Just mode' -> handleMode node mode'
             Nothing    -> Node pos ty <$> sequence subs
 
+    handleMode
+      :: (MonadWriter [ScanError] m, MonadState IgnoreMode m)
+      => Node
+      -> IgnoreMode
+      -> m Node
+    handleMode node@(Node pos _ _)= \case
+      None   -> do
+        let unrecognised = fromMaybe ""
+              $ safeHead . drop 1 . words =<< getXrefcheckContent node
```

Comment:
It seems this is duplicating some of the work in `getIgnoreMode`...

Thinking about it a bit more, I think the issue is that we're using `None` to represent two distinct things:

1. In ``action `runStateT` None`` (line 147), `None` means no "ignore mode" is currently set
2. In `getIgnoreMode / textToMode`, `None` means we found an unrecognised "ignore mode"

I think we should add another `IgnoreMode` constructor like `UnknownMode Text` for the 2nd use case.
Then, here in `handleMode`, you could simply pattern match on `UnknownMode` to obtain the name of the unrecognised mode, no need to parse it again.

Reply:
Actually, adding another constructor might not be the right thing to do either. It wouldn't make sense to do `put (UnknownMode "x")`.

Instead, I think it would make more sense to change `getIgnoreMode`. Instead of returning `Maybe IgnoreMode`, it should return something like:

```hs
data GetIgnoreMode
  = NoMode
  | ValidMode IgnoreMode
  | InvalidMode Text
```

Reply:
One more thing: with the change I proposed above, I think the `None` constructor no longer makes sense.

I.e., it wouldn't make sense for `getIgnoreMode` to return `ValidMode None`. That would be an invalid state, and according to the "make illegal states unrepresentable" mantra, the types shouldn't allow this to even be expressed.

I think we should do this:
* `IgnoreMode` should be just `Link`, `Paragraph` and `File`
* `ignoreMode` returns `NotAnAnnotation`, `ValidMode IgnoreMode` or `InvalidMode Text`
* In the `remove` helper function, we change from `MonadState IgnoreMode` to `MonadState (Maybe IgnoreMode)`, where `Nothing` represents no mode is currently set.

This is turning into a lot of work, so let's not do this in this PR, leave this part as it is. Can you please write a new follow-up issue for refactoring `IgnoreMode`?